beacon_node/lighthouse_network/src/rpc/mod.rs

```diff
@@ -117,10 +121,15 @@ pub struct NetworkParams {
 /// Implements the libp2p `NetworkBehaviour` trait and therefore manages network-level
 /// logic.
 pub struct RPC<Id: ReqId, E: EthSpec> {
-    /// Rate limiter
-    limiter: Option<RateLimiter>,
+    /// Rate limiter for our responses. This is shared with RPCHandlers.
+    response_limiter: Option<Arc<Mutex<RateLimiter>>>,
     /// Rate limiter for our own requests.
-    self_limiter: Option<SelfRateLimiter<Id, E>>,
+    outbound_request_limiter: Option<SelfRateLimiter<Id, E>>,
+    /// Limiter for inbound requests, which checks the request size.
+    inbound_request_size_limiter: Option<RequestSizeLimiter>,
```

Comment:
That might be better in terms of cleaner code. The reason why I introduced `RequestSizeLimiter` is that `RateLimiter` is shared between behaviour and connection handler as `Arc<Mutex<RateLimiter>>`. Locking time will be increased if we move `RequestSizeLimiter` inside `RateLimiter` and call the (shared) limiter in `on_connection_handler_event`. 

For reference, here is the reason why I introduced `RequestSizeLimiter`, quoted from [the PR comment](https://github.com/sigp/lighthouse/pull/5923#issue-2351213755):

> This request size check is also performed in `Before` using the `Limiter`, but in this PR, I introduced `RequestSizeLimiter` to handle this. Unlike the `Limiter`, `RequestSizeLimiter` is dedicated to perform this check.
>
> The reasons why I introduced `RequestSizeLimiter` are:
>
> - In `After`, the rate limiter is shared between the behaviour and the handler (`Arc<Mutex<RateLimiter>>>`). (This is detailed at the next sentence)
> - The request size check does not heavily depend on the rate-limiting logic, so it can be separated with minimal code duplication.
> - The request size check is performed on the behaviour side. By separating the request size check from the rate limiter, we can reduce the locking of the rate limiter.

Reply:
Suggestion from pawan about `RequestSizeLimiter`: https://github.com/sigp/lighthouse/pull/5923#discussion_r1689948084

I removed the `RequestSizeLimiter` and added `is_request_size_too_large()` instead, which checks if the count of requested blocks/blobs is within the number defined in the spec. That looks much simpler. 🙂  https://github.com/sigp/lighthouse/pull/5923/commits/5a9237fdc6a5ad11281ec7c36b2b7bdfdeba7ed4